ocaml/xapi/xapi_ha_vm_failover.ml

```diff
@@ -175,60 +203,429 @@ let order_f (_, vm_rec) =
 
 let ( $ ) x y = x y
 
-(*****************************************************************************************************)
-(* Planning code follows                                                                             *)
-
 (* Compute the total memory required of a VM (Running or not) *)
 let total_memory_of_vm ~__context policy snapshot =
   let main, shadow =
     Memory_check.vm_compute_start_memory ~__context ~policy snapshot
   in
   Int64.add main shadow
 
-(** Return a VM -> Host plan for the Host.evacuate code. We assume the VMs are all agile. The returned plan may
-    be incomplete if there was not enough memory. *)
-let compute_evacuation_plan ~__context total_hosts remaining_hosts
-    vms_and_snapshots =
-  let hosts =
-    List.map
-      (fun host ->
-        ( host
-        , Memory_check.host_compute_free_memory_with_maximum_compression
-            ~__context ~host None
+(* Used to sort pairs into increasing order of their second component *)
+let compare_snd (_, a) (_, b) = compare a b
+
+let host_free_memory ~__context ~host =
+  Memory_check.host_compute_free_memory_with_maximum_compression ~__context
+    ~host None
+
+let vm_memory ~__context snapshot =
+  let policy =
+    match Helpers.check_domain_type snapshot.API.vM_domain_type with
+    | `hvm | `pv ->
+        Memory_check.Dynamic_min
+    | `pv_in_pvh | `pvh ->
+        Memory_check.Static_max
+  in
+  total_memory_of_vm ~__context policy snapshot
+
+type anti_affinity_pool_state = {
+    host_size_map: int64 Xapi_vm_helpers.HostMap.t
+  ; grp_host_vm_cnt_map: int Xapi_vm_helpers.HostMap.t VMGroupMap.t
+  ; grp_hosts_set: HostsSet.t VMGroupMap.t
+}
```

Comment:
Could it be `plan state`? And add `plan` into it as well?
So it could be module like:
```
module PlanState = struct
  type t = {
    free_mems: int64 HostMap.t
  ; vm_cnts: int HostMap.t VMGroupMap.t
  ; hosts: HostsSet.t VMGroupMap.t
  ; plan:  [`VM] Ref.t * [`host] Ref.t list
  }

  let update t group h vm_size = ...
end
end
```

Reply:
I would think it more suitable as a `spread_evenly_pool_state` and `no_breach_pool_state`, for a "plan state", what I will think of would be: what is the state of the planning now: if it is in the "step 1 - spread evenly plan", or "step 1 - binpack plan after spread evenly plan" or "step 2", "step 3".